app/com/linkedin/drelephant/tuning/hbt/MRApplicationData.java

```diff
@@ -0,0 +1,742 @@
+    validHeuristic.add(MAPPER_MEMORY_HEURISTIC);
+    validHeuristic.add(MAPPER_SPILL_HEURISTIC);
+    validHeuristic.add(REDUCER_MEMORY_HEURISTIC);
+    MRConstant.MRConfigurationBuilder.buildConfigurations(ElephantContext.instance().getAutoTuningConf());
+  }
+
+  /**
+   *
+   * @param result : Appresults : It contains application data.
+   * @param appliedParameter :  Last applied parameter.
+   * It process the application for suggested parameter
+   */
+  MRApplicationData(AppResult result, Map<String, Double> appliedParameter) {
+    logger.info(" Process for following app id " + result.id);
+    this.applicationID = result.id;
+    this._result = result;
+    this.suggestedParameter = new HashMap<String, Double>();
+    this.failedHeuristics = new HashMap<String, AppHeuristicResult>();
+    this.appliedParameter = appliedParameter;
+    this.counterValues = new HashMap<String, Double>();
+    processForSuggestedParameter();
+    postFillSuggestedParameterWithAppliedParameter();
+  }
+
+  public Map<String, Double> getCounterValues() {
+    return this.counterValues;
+  }
+
+  public String getApplicationID() {
+    return this.applicationID;
+  }
+
+  /**
+   *Suggested parameter will suggest the parameters based on current application , but
+   *if the no parameters are suggested , then fill the suggested parameter with previous value.
+   * If no parameters are suggested , then it means heuristics are working fine
+   * with previous applied parameters and hence those parameters will again be suggested
+   * for this application
+   * Fill the Mapper parameter , with previous values ,if not suggested
+   * Fill reducer values if not suggested , only in the case reducer is there in the application
+   */
+  private void postFillSuggestedParameterWithAppliedParameter() {
+    int numberOfReducer = getPreviousNumberOfReducer();
+    postFillReducerParameter(numberOfReducer);
+    postFillMapperParameter();
+    if (debugEnabled) {
+      for (String parameterName : suggestedParameter.keySet()) {
+        logger.debug(" Suggested Parameter Test " + parameterName + "\t" + suggestedParameter.get(parameterName));
+      }
+    }
+  }
+
+  /**
+   * This method will fill the reducer releated parameters from applied parameters ,
+   * if they are not suggested .
+   * @param numberOfReducer : Number of reducer of the application
+   */
+  private void postFillReducerParameter(int numberOfReducer) {
+    if (numberOfReducer > 0) {
+      if (!suggestedParameter.containsKey(REDUCER_MEMORY_HADOOP_CONF.getValue())) {
+        suggestedParameter.put(REDUCER_MEMORY_HADOOP_CONF.getValue(),
+            appliedParameter.get(REDUCER_MEMORY_HEURISTICS_CONF.getValue()));
+      }
+      if (!suggestedParameter.containsKey(REDUCER_HEAP_HADOOP_CONF.getValue())) {
+        suggestedParameter.put(REDUCER_HEAP_HADOOP_CONF.getValue(),
+            appliedParameter.get(REDUCER_HEAP_HEURISTICS_CONF.getValue()));
+      }
+    } else {
```

Comment:
Idea is same only , not set any value if algorithm is not suggesting any value for that application . If application one suggested value for parameter P1 and application 2 has not suggested value that means its should be the default value (But the same is not true in case of reducers ) . Since if there is no reducer for the application ,we cannot set default value for the same .

Reply:
If there are no reducer there is no harm in setting default reducer value. But if we set reducer to 0 and then if there is code changes and suddenly reducers are introduced, we will keep setting 0.

Reply:
ok. This suggested parameter is at the application level and not at job level . So if there is no reducer in the application , we should set reducer memory to zero and not to default value . Because , if we set the default value ,at the application level , then these default value may be the final max value suggested at the job level , which is not correct .  I'll explain you the same offline also